CMakeLists.txt

```diff
@@ -1,31 +1,75 @@
-cmake_minimum_required(VERSION 3.8)
-project(snmalloc C CXX)
+cmake_minimum_required(VERSION 3.14)
+project(snmalloc CXX)
 
 if (NOT CMAKE_BUILD_TYPE)
   message(STATUS "No build type selected, default to: Release")
   set(CMAKE_BUILD_TYPE "Release")
 endif()
 
 include(CheckCXXCompilerFlag)
-include(CheckCSourceCompiles)
+include(CheckCXXSourceCompiles)
+include(CMakeDependentOption)
 
+option(SNMALLOC_HEADER_ONLY_LIBRARY "Use snmalloc has a header-only library" OFF)
+# Options that apply globally
 option(USE_SNMALLOC_STATS "Track allocation stats" OFF)
 option(SNMALLOC_CI_BUILD "Disable features not sensible for CI" OFF)
-option(EXPOSE_EXTERNAL_PAGEMAP "Expose the global pagemap" OFF)
-option(EXPOSE_EXTERNAL_RESERVE "Expose an interface to reserve memory using the default memory provider" OFF)
-option(SNMALLOC_RUST_SUPPORT "Build static library for rust" OFF)
-option(SNMALLOC_STATIC_LIBRARY   "Build static libraries" ON)
 option(SNMALLOC_QEMU_WORKAROUND "Disable using madvise(DONT_NEED) to zero memory on Linux" Off)
-option(SNMALLOC_OPTIMISE_FOR_CURRENT_MACHINE "Compile for current machine architecture" Off)
-set(SNMALLOC_STATIC_LIBRARY_PREFIX "sn_" CACHE STRING "Static library function prefix")
 option(SNMALLOC_USE_CXX17 "Build as C++17 for legacy support." OFF)
-option(SNMALLOC_USE_PTHREAD_DESTRUCTORS "Build using pthread destructors. Reduces the system dependencies, but may interact badly with C++ on some platforms." OFF)
+# Options that apply only if we're not building the header-only library
+cmake_dependent_option(SNMALLOC_RUST_SUPPORT "Build static library for rust" OFF "NOT SNMALLOC_HEADER_ONLY_LIBRARY" OFF)
+cmake_dependent_option(SNMALLOC_STATIC_LIBRARY "Build static libraries" ON "NOT SNMALLOC_STATIC_LIBRARY" OFF)
+cmake_dependent_option(SNMALLOC_OPTIMISE_FOR_CURRENT_MACHINE "Compile for current machine architecture" Off "NOT SNMALLOC_STATIC_LIBRARY" OFF)
+if (NOT SNMALLOC_HEADER_ONLY_LIBRARY)
+  # Pick a sensible default for the thread cleanup mechanism
+  if (${CMAKE_SYSTEM_NAME} STREQUAL FreeBSD)
+    set(SNMALLOC_CLEANUP_DEFAULT THREAD_CLEANUP)
+  elseif (UNIX)
```

Comment:
This doesn't work on Mac as the interaction with pthread and C++ destructors is bad.  It appeared the C++ TLS state had been zeroed before the pthread thread destructor was called. and then bad things happened.